ctapipe/image/extractor.py

```diff
@@ -1285,3 +1290,373 @@ def __call__(
+        "1: local pixel counts as much as any neighbor)",
+    ).tag(config=True)
+
+    effective_time_profile_std = FloatTelescopeParameter(
+        default_value=2.0,
+        help="Effective Cherenkov time profile std. dev. (in nanoseconds) to "
+        "assume for calculating the gain correction",
+    ).tag(config=True, min=0)
+
+    neighbour_sum_clipping = FloatTelescopeParameter(
+        default_value=5.0,
+        help="(Soft) clipping level of a pixel's contribution to a neighbour sum "
+        "(set to 0 or inf to disable clipping)",
+    ).tag(config=True)
+
+    leading_edge_timing = BoolTelescopeParameter(
+        default_value=True, help="Calculate leading edge time instead of peak time"
+    ).tag(config=False)
+
+    leading_edge_rel_descend_limit = FloatTelescopeParameter(
+        default_value=0.05,
+        help="Fraction of the peak value down to which samples are accumulated "
+        "in the leading edge centroid calculation",
+    ).tag(config=True, min=0.0, max=1.0)
+
+    def __init__(self, subarray, **kwargs):
+        super().__init__(subarray=subarray, **kwargs)
+
+        self.sampling_rate_ghz = {
+            tel_id: telescope.camera.readout.sampling_rate.to_value("GHz")
+            for tel_id, telescope in subarray.tel.items()
+        }
+
+        def time_profile_pdf_gen(std_dev: float):
+            if std_dev == 0:
+                return None
+            return scipy.stats.norm(0.0, std_dev).pdf
+
+        self.deconvolution_pars = {
+            tel_id: deconvolution_parameters(
+                tel.camera,
+                self.upsampling.tel[tel_id],
+                self.window_width.tel[tel_id],
+                self.window_shift.tel[tel_id],
+                self.leading_edge_timing.tel[tel_id],
+                self.leading_edge_rel_descend_limit.tel[tel_id],
+                time_profile_pdf_gen(self.effective_time_profile_std.tel[tel_id]),
+            )
+            for tel_id, tel in subarray.tel.items()
+        }
+
+    @staticmethod
+    def clip(x, lo=0.0, hi=np.inf):
+        """Applies soft clipping to ±1 and then hard clipping to (lo, hi)."""
+        return np.clip(x / (1.0 + np.abs(x)), lo, hi)
+
+    def __call__(
+        self, waveforms, tel_id, selected_gain_channel, broken_pixels
+    ) -> DL1CameraContainer:
+        upsampling = self.upsampling.tel[tel_id]
+        integration_window_width = self.window_width.tel[tel_id]
+        integration_window_shift = self.window_shift.tel[tel_id]
+        neighbour_sum_clipping = self.neighbour_sum_clipping.tel[tel_id]
+        leading_edge_timing = self.leading_edge_timing.tel[tel_id]
+        leading_edge_rel_descend_limit = self.leading_edge_rel_descend_limit.tel[tel_id]
+
+        pzs, gains, shifts, pz2ds = self.deconvolution_pars[tel_id]
+        pz, gain, shift, pz2d = pzs[0], gains[0], shifts[0], pz2ds[0]
+
+        if leading_edge_timing:
+            d_waveforms = deconvolve(waveforms, 0.0, upsampling, 1)
```

Comment:
This is calculated up here but only used much further down in another if statement. 

You should move it down to the place where it is actually used.